bip-xxxx.mediawiki

```diff
@@ -0,0 +1,303 @@
+
+To prevent man-in-the-middle attacks, the sender must ensure either that `https` is used, or that the domain ends with `.onion` to support receivers over a Tor Hidden Service.
+
+===Receiver's well known errors===
+
+If for some reason the receiver is unable to create a payjoin proposal, it will reply with a HTTP code different than 200.
+The receiver is not constrained to specific set of errors, some are specified in this proposal.
+
+The errors have the following format:
+<pre>
+{
+    "errorCode": "leaking-data",
+    "message": "Key path information or GlobalXPubs should not be included in the original PSBT."
+}
+</pre>
+
+The well-known error codes are:
+{| class="wikitable"
+!Error code
+!Meaning
+|-
+|leaking-data
+|Key path information or GlobalXPubs should not be included in the original PSBT.
+|-
+|psbt-not-finalized
+|The original PSBT must be finalized.
+|-
+|unavailable
+|The payjoin endpoint is not available for now.
+|-
+|out-of-utxos
+|The receiver does not have any UTXO to contribute in a payjoin proposal.
+|-
+|not-enough-money
+|The receiver added some inputs but could not bump the fee of the payjoin proposal.
+|-
+|insane-psbt
+|Some consistency check on the PSBT failed.
+|-
+|version-unsupported
+|This version of payjoin is not supported.
+|-
+|need-utxo-information
+|The witness UTXO or non witness UTXO is missing
+|}
+
+The receiver is allowed to return implementation specific errors which may assist the sender to diagnose any issue.
+
+However, it is important that error codes that are not well-known and that the message do not appear on the sender's software user interface.
+Such error code or message could be used maliciously to fish a non technical user.
+Instead those error or message can only appear in debug logs.
+
+It is advise to hard code the description of the error code into the sender's software.
+
+===Sender's payjoin proposal checklist===
+
+The sender should check the payjoin proposal before signing it to prevent a malicious receiver from stealing money.
+
+* Check that all the spent outpoints in the original PSBT still exists in the coinjoin PSBT.
+* Check that all the spent outpoints in the original PSBT does not have any partial signature.
+* If the sender is not using inputs with mixed types, check that the receiver inputs type match the inputs type of the sender. (ie. both using P2SH-P2WPKH or both using P2WPKH)
+* Check that any inputs added by the receiver are finalized.
+* Check that the transaction version, and nLockTime are unchanged.
+* Check that the sender's inputs' sequence numbers are unchanged.
+* Check that the sent amount in the payjoin proposal is less or equal to the sent amount of the original transaction.
+
+If the sent amount in the payjoin proposal is above the amount sent in the original PSBT
+* Check that the additional paid amount has been paid to fee.
+* Check that the additional paid amount is not more than twice the original fee 
```

Comment:
I had the same thought. Why not require the receiver to bump any fees required?

IMO, if the sender has already chosen a suitable fee rate and is expecting a certain amount of change, the receiver should not be changing this.

If the receiver wants to add inputs/outputs that's fine but they should cover all fees required and ensure they maintain an equivalent fee rate.

Allowing the receiver to change these values can lead to all sorts of strange UX issues that are problematic to solve on the wallet end.

Reply:
What would be more reasonable here to state in terms of being cautious of fee increases?

Reply:
I think change:

>- Check that the sent amount in the payjoin proposal is less or equal to the sent amount of the original transaction.
>
>If the sent amount in the payjoin proposal is above the amount sent in the original PSBT
>- Check that the additional paid amount has been paid to fee.
>- Check that the additional paid amount is not more than twice the original fee
>- Check that the additional paid amount has only been used to pay the receiver's additional inputs.

to just:
>- Check that the sent amount in the payjoin proposal is equal to the sent amount of the original transaction.

Reply:
Many wallets ask a user to confirm the details of a transaction before sending it. This normally includes fee rate and spending amount. It's pretty confusing if a user confirms this, does a PayJoin, and ends up broadcasting a TX that actually spends a different amount with a different fee rate.

Yes, the wallet could prompt the user for confirmation but that leads to other UX issues and weird edge cases.

If the wallet prompts the user if they want to accept on fee rate / change amount change, the original non-payjoin TX could already be broadcast by the receiver before they confirm. This would be very confusing.

If we don't prompt the user they could lose unfair amounts of money or the TX could take an unreasonable length of time to confirm.

Maybe sender wanted a specific fee rate, because they want the change UTXOs in a reasonable amount of time.
Maybe sender wanted a specific amount of change.

The receiver shouldn't be changing these. The receiver can add whatever they want, and the sender shouldn't care, as long as their change/fee rate values aren't changed.

Reply:
It also seems like the receive has more to gain from the PayJoin than the sender.

If they use the senders change to bump the fee they are basically helping theirselves to their customers money to perform their own payment/consolidation.

Another reason why I think the receiver should be covering all extra fees.

Reply:
An example of how this could be pretty unfair in the current fee market, not even a future high fee market:

1. A customer is purchasing a subscription from a merchant.
2. Subscription is low cost, let's say $5 equivalent.
3. User needs to pay a TX fee of $1 equiv to get confirmed promptly.
4. User accepts this fee because they need access to the service now.
5. User does a PJ transaction with merchant.
6. Merchant batches in a load of extra TXs, uses the full allowable doulbe fee limit and bumps TX fee to $2.

In that scenario the user only agreed to pay $5 to the merchant and $1 to the network, yet the merchant was able to help themselves to an extra 20% of the originally agreed payment amount to batch their own transactions.

Reply:
> If we don't prompt the sender, they could lose unfair amounts of money or the TX could take an unreasonable length of time to confirm.

I think it's important to be clear which part of the proposal you're critiquing; this point I quoted above, was already precisely addressed by the existing text/protocol (the fee rate was kept unchanged).

Your main point, though, I agree, it's important to attribute costs fairly, *but* it does seem a little tricky, as it violates what I suspect is a rather key expectation in merchant wallet code: a transaction paying an invoice fully and precisely pays the expected amount.

Reply:
@lukechilds the problem is that the receiver might not be too happy to activate payjoin if that cost him more money.

The initial reason of this limit was to put a limit on how many inputs the merchant can add.

I think the correct way to deal with this issue is to ask to the sender in the UX: Payjoin is costing an additional X amount. Do you wish to proceed?

Reply:
Does the spec currently say the fee rate cannot be changed?

I see it says:

>they need to make sure the fee is increased such that the rate does not drop below 1 satoshi per virtual byte.

But that doesn't require that the fee rate is unchanged, just that it isn't dropped below 1sat/vbyte.

But I have also noticed this:

>===Receiver does not need to be a full node===
>Because the receiver needs to bump the fee to keep the same fee rate as the original PSBT, it needs the input's UTXO information to know what is the original fee rate. Without PSBT, light wallets like Wasabi Wallet would not be able to receive a payjoin transaction.

Which implies that somewhere else the proposal is enforcing that. Is this the only place where it's stated the fee rate cannot be changed, or have I missed it mentioned somewhere else too?

>it violates what I suspect is a rather key expectation in merchant wallet code: a transaction paying an invoice fully and precisely pays the expected amount.

I'm not sure I understand this point, are you able to clarify? How does covering extra fee costs with the merchant input change this?

Reply:
@NicolasDorier sorry, I've only just seen your response.

>the problem is that the receiver might not be too happy to activate payjoin if that cost him more money.

But it should be cheaper to batch their pending transactions in the payjoin than to create a new transaction, plus they get to consolidate the payment UTXO.

The incentive is still there but it is not as strong as before. I think this is fair though, it was only incentivised more before because the sender was covering all the costs.

>I think the correct way to deal with this issue is to ask to the sender in the UX: Payjoin is costing an additional X amount. Do you wish to proceed?

This could lead to the potentially confusing scenario I raised here: https://github.com/NicolasDorier/bips/pull/3#discussion_r422867485.

The non-pj TX could be broadcast by the receiver before the sender has confirmed the pj TX resulting in confusing behaviour.

Reply:
@lukechilds this is difficult point here. Because I expect that if we tell our merchants that they will in general get less money, then he might not be open to activate payjoin.

As @AdamISZ  said, it is true that receiver save money since they consolidate now 2 UTXO into 1 UTXO, saving 1 input for the next transaction they make so it would make it logical they pay for it.
However, from a merchant perspective, proper accounting is already complicated thing. If now they start having to say to the taxman "The rate was X, we asked 1BTC for it, but we received 0.99999BTC because of the added fee in payjoin", this become fairly complex to handle and counter intuitive: "I asked the customer to pay for 1 BTC not for 0.999999 BTC".

About the approval timeout, I think this can be shown to the sender. Showing a timer of 2min to chose what he wants.

I was thinking about another potential solution on the receiver side, but I need to test for it:
If the original PSBT does not have a round fee, and that we are above the min relay fee, we can potentially not need of bumping fee.
That said, if RBF, it is possible that the original PSBT ends up replacing the payjoin transaction if the fee rate delta is big enough. (I don't remember how much)
If not RBF, I think it is quite safe.

Reply:
> Which implies that somewhere else the proposal is requiring that the fee rate cannot be changed. Is this the only place where it's stated the fee rate cannot be changed, or have I missed it mentioned somewhere else too?

I *think* you are right: I originally wrote an implementation based on the btcpayserver doc - see "Step 3" in this section: https://github.com/btcpayserver/btcpayserver-doc/blob/master/Payjoin-spec.md#difference-with-bustapay .. where it is specified, I think it is not specified here? So, if so, it should be.

Reply:
> I'm not sure I understand this point, are you able to clarify? How does covering extra fee costs with the merchant input change this?

I meant it at the higher, business logic level, independent of tx construction details: merchants today do not bear any cost from bitcoin network tx fees. So if they invoice 0.1BTC they are paid 0.1BTC exactly. If the sender/customer pays all of the extra fee cost of Payjoin (typically, the extra fee for 1 input), that rule is still in place.

Reply:
@AdamISZ yes, I did not specified it, because I was thinking: If the fee rate is different, if the delta is above the min delta for transaction replacement. (or that RBF is not opt-in) then it is safe to have a different, lower, fee rate. This should not be done if the fee rate is round though for privacy purpose.

Reply:
Ah, yes ok, I understand regarding accounting/invoicing complications on the receiver end now.

>Also, the sender would not really complain about not paying more, so the receiver is free to do whatever he wants.

But they may complain about their fee rate being reduced. If the sender chose a specific fee rate because it's important that their transaction is confirmed within a certain time window, it may be quite inconvenient if the payjoin tx has a reduced fee rate.

I think it should be enforced in the spec that the payjoin tx fee rate is more than or equal to the original tx fee rate.

Reply:
This point continues to be very tricky. We seem to need to consider:
* lowering fee rate could impact sender negatively (conf wait time), not desirable
* raising fee rate without receiver contributing could lead to extra sender cost beyond what is acceptable, not desirable
* using a fee rate or fee rate distribution different from what the sender wallet usually uses could watermark this as payjoin, not desirable
* maintaining the same fee rate but attributing the fair cost to the receiver, by reducing the payment amount, could screw up the merchant's accounting, not desirable
* changes in fee rate could change the situation w.r.t. RBF in a way that is ... not desirable

What is desirable? :)
I tend to lead towards a tolerance approach as per the original description ("reduction in change not > 1 x original tx fee"), but with an additional check of *percentage* change as well. And it only being a broad recommendation; any individual wallet could safely reject any payjoin option at any time because they don't like the fee.

Reply:
> but with an additional check of percentage change as well. 

This is the `Check that the additional paid amount is not more than twice the original fee ` part. Basically it allows a 100% percent change max. This is not fool proof either as high fee can make it unacceptable.

Maybe we should just remove the percent change max, and let sender the freedom to deal as they want with it? If the fee is almost the same, the sender should not care about validating the payjoin. However, if the fee is very different, the sender's UI might want to show a proper confirmation step to the user. 

Given how pain in the ass it is to implement, I guess a good temporary way for senders would be to enforce in software this max limit.

Another thing we can do is also to add an `optional parameter` that the sender can specify to the receiver "maxbumpfee". For example "maxfeecontribution=1000" would tell to the receiver that the sender wants to pay up to 1000 satoshi of fee. Above this, the receiver could still pay from his own pocket.

Reply:
>Another thing we can do is also to add an optional parameter that the sender can specify to the receiver "maxfeecontribution". For example "maxfeecontribution=1000" would tell to the receiver that the sender wants to pay up to 1000 satoshi of fee. Above this, the receiver could still pay from his own pocket.

Yeah, I think this is a reasonable solution.

It could just be a global setting in the wallet that defaults to some sane value so no need to configure it each time and no need to manually set it. Then users can set a new value if desired or override on a per payment basis.

It solves the UX issue of requiring a second user confirmation as you said. It also allows a wallet to show the max possible potential fee in the initial confirmation dialog to the user too.

Reply:
> This is the `Check that the additional paid amount is not more than twice the original fee ` part. Basically it allows a 100% percent change max. This is not fool proof either as high fee can make it unacceptable.

Sorry if it wasn't clear. I meant percentage *of the payment amount*, not percentage of the existing fee. And this would be separate to what I am counting as a non-percentage "absolute" measure - the existing fee in satoshis (100% *of fee* as a number of sats).

Reply:
I wrote https://github.com/NicolasDorier/bips/pull/3/commits/38cfb3f727945abe25e1675a0303e691ec09b7c2 @AdamISZ @lukechilds? this would allow the sender to decide by himself what he considers reasonable.

Reply:
I think that's good.

Maybe reword to:

```
* <code>maxfeebumpcontribution=</code>, an integer defining the maximum amount in satoshis that the sender is willing to contribute towards fees for the additional inputs. To be ignored if set to less than zero. (default: -1)
```